examples/react-tic-tac-toe/src/ticTacToe.machine.ts

```diff
@@ -0,0 +1,126 @@
+import { createMachine } from 'xstate';
+import {
+  createModel,
+  ModelContextFrom,
+  ModelEventsFrom
+} from 'xstate/lib/model';
+
+type Player = 'x' | 'o';
+
+const model = createModel(
+  {
+    board: Array(9).fill(null) as Array<Player | null>,
+    moves: 0,
+    player: 'x' as Player,
+    winner: undefined as Player | undefined
+  },
+  {
+    events: {
+      PLAY: (value: number) => ({ value }),
+      RESET: () => ({})
+    }
+  }
+);
+
+const isValidMove = (
+  ctx: ModelContextFrom<typeof model>,
+  e: ModelEventsFrom<typeof model> & { type: 'PLAY' }
+) => {
+  return ctx.board[e.value] === null;
+};
+
+function checkWin(ctx: ModelContextFrom<typeof model>) {
+  const { board } = ctx;
+  const winningLines = [
+    [0, 1, 2],
+    [3, 4, 5],
+    [6, 7, 8],
+    [0, 3, 6],
+    [1, 4, 7],
+    [2, 5, 8],
+    [0, 4, 8],
+    [2, 4, 6]
+  ];
+
+  for (let line of winningLines) {
+    const xWon = line.every((index) => {
+      return board[index] === 'x';
+    });
+
+    if (xWon) {
+      return true;
+    }
+
+    const oWon = line.every((index) => {
+      return board[index] === 'o';
+    });
+
+    if (oWon) {
+      return true;
+    }
+  }
+}
+
+function checkDraw(ctx: ModelContextFrom<typeof model>) {
+  return ctx.moves === 9;
+}
+
+export const ticTacToeMachine = createMachine<typeof model>(
+  {
+    initial: 'playing',
+    context: model.initialContext,
+    states: {
+      playing: {
+        always: [
+          { target: 'winner', cond: 'checkWin' },
+          { target: 'draw', cond: 'checkDraw' }
+        ],
+        on: {
+          PLAY: [
+            {
+              target: 'playing',
+              cond: 'isValidMove',
+              actions: 'updateBoard'
+            }
+          ],
+          RESET: undefined // don't allow resetting while playing
```

Comment:
it feels like preventing transitions like this should be used rarely - if you don't look close enough at the visualization this is like a piece of hidden information because if you are "zoomed out" you might think that `RESET` can happen at any time.

feels like maybe `winner` and `draw` could be grouped together in like a `gameEnd` state and that state could handle `RESET` (to avoid declaring this transition on both `winner` and `draw` - which would also be a valid approach to this problem)